File: library/src/main/java/androidz/util/DpUtil.java
package androidz.util;

import android.content.Context;

public class DpUtil {

    public static int dp2px(Context context, float dp) {
        float scale = context.getResources().getDisplayMetrics().density;
        return (int) (dp * scale + 0.5f);
    }

    public static float dp2pxF(Context context, float dp) {
        float scale = context.getResources().getDisplayMetrics().density;
        return dp * scale;
    }

    public static int px2dp(Context context, float px) {
        float scale = context.getResources().getDisplayMetrics().density;
        return (int) (px / scale + 0.5f);
    }

    public static float px2dpF(Context context, float px) {
        float scale = context.getResources().getDisplayMetrics().density;
        return px / scale;
    }


    public static int px2sp(Context context, float px) {
        float scale = context.getResources().getDisplayMetrics().scaledDensity;
        return (int) (px / scale + 0.5f);
    }

    public static float px2spF(Context context, float px) {
        float scale = context.getResources().getDisplayMetrics().scaledDensity;
        return px / scale;
    }

    public static int sp2px(Context context, float sp) {
        float scale = context.getResources().getDisplayMetrics().scaledDensity;
        return (int) (sp * scale + 0.5f);
    }

    public static float sp2pxF(Context context, float sp) {
        float scale = context.getResources().getDisplayMetrics().scaledDensity;
        return sp * scale;
    }

}

File: library/src/main/java/androidz/util/DpUtil.kt
package androidz.util

import android.content.Context


object DpUtil {

    @JvmStatic
    fun dp2px(context: Context, dp: Float): Int {
        val scale = context.resources.displayMetrics.density
        return (dp * scale + 0.5f).toInt()
    }

    @JvmStatic
    fun dp2pxF(context: Context, dp: Float): Float {
        val scale = context.resources.displayMetrics.density
        return dp * scale
    }

    @JvmStatic
    fun px2dp(context: Context, px: Float): Int {
        val scale = context.resources.displayMetrics.density
        return (px / scale + 0.5f).toInt()
    }

    @JvmStatic
    fun px2dpF(context: Context, px: Float): Float {
        val scale = context.resources.displayMetrics.density
        return px / scale
    }

    @JvmStatic
    fun px2sp(context: Context, px: Float): Int {
        val scale = context.resources.displayMetrics.scaledDensity
        return (px / scale + 0.5f).toInt()
    }

    @JvmStatic
    fun px2spF(context: Context, px: Float): Float {
        val scale = context.resources.displayMetrics.scaledDensity
        return px / scale
    }

    @JvmStatic
    fun sp2px(context: Context, sp: Float): Int {
        val scale = context.resources.displayMetrics.scaledDensity
        return (sp * scale + 0.5f).toInt()
    }

    @JvmStatic
    fun sp2pxF(context: Context, sp: Float): Float {
        val scale = context.resources.displayMetrics.scaledDensity
        return sp * scale
    }
}
